### PR TITLE
Resolved problems and tests

### PR DESCRIPTION
**Pull request by MateaAndrei**

### CODE REVIEW

File: Problem1.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+def SumOfEven numbers =[]
+sum=0
+
+    if numbers.nil?
+        raise ArgumentError, 'Argument is null' 
+    end  
+   
+    numbers.each do|number|
+        if !number.is_a? Integer
+           raise ArgumentError, 'Argument is not numeric'
```
Comment (alexandru-calinoiu) [inline Problem1.rb]: The problem statement does not mention anything else then numbers in the array, I don't this brings any value, in fact in makes the code harder to read.

Using negation here make the code harder to reason about, there are better ways to do this in ruby, can you think of one?

File: Problem1.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        if !number.is_a? Integer
+           raise ArgumentError, 'Argument is not numeric' 
+        end
+
+        if number.even?
+           sum +=number
+         end  
+     end      
+    
+    return sum
```
Comment (alexandru-calinoiu) [inline Problem1.rb]: The return keyword is not needed.

File: Problem1.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# sum of evem numbers from integer array
+
+def SumOfEven numbers =[]
```
Comment (alexandru-calinoiu) [inline Problem1.rb]: A very procedural way of writing this, but ruby is much more expressive and simple, can you rewrite this in one line?

File: Problem2.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# 2 largest numbers from integer array
+
+
+def TwoBiggestNumbers numbers
+
+
+
+   if numbers.nil?
```
Comment (alexandru-calinoiu) [inline Problem2.rb]: Why will you send the numbers nil? again the problem statement is clear about this, you should not need this check.

File: Problem2.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+
+
+   if numbers.nil?
+      raise ArgumentError, 'Argument is null' 
+   end   
+   if numbers.count < 2
+      raise ArgumentError, 'At least 2 elements in the array are required' 
+   end
+   numberA= numberB= numbers[0]
```
Comment (alexandru-calinoiu) [inline Problem2.rb]: What do `numberA` and `numberB` mean? 

The problem is about finding the 2 largest numbers, can you rename your variables so they make better sens for someone reading this code?

File: Problem3.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+     raise ArgumentError, 'Argument is null' 
+  end 
+
+  numbers.each   do |number|
+
+    if !number.is_a? Integer
+      raise ArgumentError, 'Argument is not numeric' 
+    end
+
+    if !number.positive?
```
Comment (alexandru-calinoiu) [inline Problem3.rb]: This is the perfect example why this checks should not exist :)

Please try to run your code with the following input [2, 1, 1].

File: Problem3.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      raise ArgumentError, 'Argument is not a positive number' 
+    end
+     
+    if numbers[number.abs-1] >0
```
Comment (alexandru-calinoiu) [inline Problem3.rb]: Even if the problem was correctly implemented this is not the right solution, for an input like [2, 1] you will fail by returning that the array has duplicates.

Please try to find another solution.

File: TestProblem1.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
+require_relative "Problem1"
+require "Test/unit"
+ 
+class TestProblem1 < Test::Unit::TestCase
+ 
+  def test_SumOfEvenIsCorrect
+
+    sum= SumOfEven(Array(-4..9))
+    assert_not_nil(sum)
```
Comment (alexandru-calinoiu) [inline TestProblem1.rb]: Tests should contain only one assertion.

File: TestProblem2.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
+require_relative "Problem2"
+require "test/unit"
+ 
+class TestProblem2 < Test::Unit::TestCase
+ 
+  def test_CorrectTwoBiggestNumbers
+
+    numbers= TwoBiggestNumbers(Array(-4..9))
```
Comment (alexandru-calinoiu) [inline TestProblem2.rb]: Cool way of generating an array :+1:

File: TestProblem2.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+ 
+class TestProblem2 < Test::Unit::TestCase
+ 
+  def test_CorrectTwoBiggestNumbers
+
+    numbers= TwoBiggestNumbers(Array(-4..9))
+    assert_not_nil(numbers)
+    assert_equal([9,8],numbers)
+    
+    numbers= TwoBiggestNumbers([3,3,3,3,3,3])
```
Comment (alexandru-calinoiu) [inline TestProblem2.rb]: Please refactor this to be 2 tests.

File: TestProblem2.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    assert_not_nil(numbers)
+    assert_equal([3,3],numbers)
+
+  end
+
+
+
+  def test_RaiseExceptionWhenArgumentIsIncorrect
+
+    assert_raise( ArgumentError ) { TwoBiggestNumbers(nil) }
```
Comment (alexandru-calinoiu) [inline TestProblem2.rb]: Please refactor this into multiple tests.

File: TestProblem3.rb [hunk 12]
```diff
@@ -0,0 +1,28 @@
+require_relative "Problem3"
+require "test/unit"
+ 
+class TestProblem3 < Test::Unit::TestCase
```
Comment (alexandru-calinoiu) [inline TestProblem3.rb:4]: Mostly same comments as in the previous tests.

**Review by alexandru-calinoiu — CHANGES_REQUESTED** [review]
This is a good start.

Apart from the code comments there are 2 more things that you should consider:
- the code looks very messy and not very ruby like please take a look at the style guide https://github.com/rubocop-hq/ruby-style-guide and adopt it for your solutions
- tests are very poorly written please consider following a pattern like arrange-act-assert http://wiki.c2.com/?ArrangeActAssert with only 1 assertion per test.

Keep going!

File: Problem3.rb (hunk 7 above)
Comment (MateaAndrei) [reply Problem3.rb]: I'm not sure i understand what you mean here. I added a test for this case and it seems to work. Can you give me more details of what you meant?

File: Problem1.rb [hunk 13]
```diff
@@ -1,22 +1,5 @@
...
-
-        if number.even?
-           sum +=number
-         end  
-     end      
-    
-    return sum
-
+def sum_of_even numbers   
+   numbers.select { |number| number.even?}.sum
```
Comment (alexandru-calinoiu) [inline Problem1.rb:4]: Much better!

To be really nitpick the style is still not perfect, there is a missing space, do you know that most modern editors have a function to automatically format your code?

File: Problem2.rb [hunk 14]
```diff
@@ -1,31 +1,6 @@
...
-      elsif number>numberB 
-         numberB=number
-      end    
-
-    end
-
-     return [numberA, numberB]    
-
+def two_biggest_numbers numbers
+   numbers.max(2)
```
Comment (alexandru-calinoiu) [inline Problem2.rb:5]: Better, 

Are you sure `max` does not sort the array in the implementation? (the request was not to sort the array)

File: Problem3.rb [hunk 15]
```diff
@@ -2,31 +2,18 @@
 include Math
 
 
-def ContainsDuplicates numbers
-
-  if numbers.nil?
-     raise ArgumentError, 'Argument is null' 
-  end 
+def contains_duplicates numbers
```
Comment (alexandru-calinoiu) [inline Problem3.rb:5]: The style :(

Running rubocop on this file you get 20 offenses:

```
C

Offenses:

duplicates.rb:1:1: C: Layout/LeadingCommentSpace: Missing space after #.
#check for duplicates from from positive integers array
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
duplicates.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
#check for duplicates from from positive integers array
^
duplicates.rb:2:1: C: Style/MixinUsage: include is used at the top level. Use inside class or module.
include Math
^^^^^^^^^^^^
duplicates.rb:4:1: C: Layout/EmptyLines: Extra blank line detected.
duplicates.rb:5:25: C: Style/MethodDefParentheses: Use def with parentheses when there are parameters.
def contains_duplicates numbers
                        ^^^^^^^
duplicates.rb:6:1: C: Layout/EmptyLinesAroundMethodBody: Extra empty line detected at method body beginning.
duplicates.rb:7:18: C: Layout/SpaceAroundOperators: Surrounding space missing for operator =.
  duplicate_found=false
                 ^
duplicates.rb:8:15: C: Layout/ExtraSpacing: Unnecessary spacing detected.
  numbers.each   do |number|
              ^^
duplicates.rb:9:26: C: Layout/SpaceAroundOperators: Surrounding space missing for operator -.
    if numbers[number.abs-1].positive?
                         ^
duplicates.rb:10:25: C: Layout/SpaceAroundOperators: Surrounding space missing for operator -.
      numbers[number.abs-1] *= -1 
                        ^
duplicates.rb:10:34: C: Layout/TrailingWhitespace: Trailing whitespace detected.
      numbers[number.abs-1] *= -1 
                                 ^
duplicates.rb:11:4: C: Layout/ElseAlignment: Align else with if.
   else
   ^^^^
duplicates.rb:12:4: C: Layout/IndentationWidth: Use 2 (not 1) spaces for indentation.
    duplicate_found = true
   ^
duplicates.rb:13:6: C: Layout/IndentationConsistency: Inconsistent indentation detected.
     break              
     ^^^^^
duplicates.rb:13:11: C: Layout/TrailingWhitespace: Trailing whitespace detected.
     break              
          ^^^^^^^^^^^^^^
duplicates.rb:15:6: C: Layout/TrailingWhitespace: Trailing whitespace detected.
  end 
     ^
duplicates.rb:16:2: C: Layout/IndentationConsistency: Inconsistent indentation detected.
 duplicate_found
 ^^^^^^^^^^^^^^^
duplicates.rb:17:1: C: Layout/TrailingWhitespace: Trailing whitespace detected.
duplicates.rb:18:4: C: Layout/TrailingWhitespace: Trailing whitespace detected.
end 
   ^
duplicates.rb:18:5: C: Layout/TrailingBlankLines: 1 trailing blank lines detected.
end  ...
    

1 file inspected, 20 offenses detected
```

File: TestProblem2.rb [hunk 16]
```diff
@@ -0,0 +1,47 @@
+require_relative "Problem2"
+require "test/unit"
+ 
+class TestProblem2 < Test::Unit::TestCase
```
Comment (alexandru-calinoiu) [inline TestProblem2.rb:4]: Better tests!

File: TestProblem2.rb [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+  
+  end
+
+  def test_correct_with_one_number_array
+
+    inputArray = [5]
+
+    numbers= two_biggest_numbers(inputArray)
+
+    assert_equal([5],numbers)
```
Comment (alexandru-calinoiu) [inline TestProblem2.rb:33]: This differs from your initial implementation, you raised an error when the array was less than 2 in length.

Is the current approach better or you initial one, please justify your answer?

**Review by alexandru-calinoiu — CHANGES_REQUESTED** [review]

File: Problem1.rb (hunk 1 above)
Comment (alexandru-calinoiu) [reply Problem1.rb]: The above question is still valid, please let me know what will be the alternative to a negating if in ruby?

File: Problem3.rb (hunk 7 above)
Comment (alexandru-calinoiu) [reply Problem3.rb]: Sorry my mistake, the code was hard to follow and I thought you missed an `.abs`

File: Problem1.rb (hunk 1 above)
Comment (MateaAndrei) [reply Problem1.rb]: if not  or unless statements

File: Problem2.rb (hunk 14 above)
Comment (MateaAndrei) [reply Problem2.rb:5]: Based on what i have searched, it doesn't look that max is sorting the array

**Comment by MateaAndrei** [comment]
>

File: TestProblem2.rb (hunk 17 above)
Comment (MateaAndrei) [reply TestProblem2.rb:33]: in the first comment from Problem1.rb you said: 
> The problem statement does not mention anything else then numbers in the array, I don't this brings any value, in fact in makes the code harder to read.<
 So i read the problems statements again and in there, there was no saying that the array should be a minimum of 2 members. I think the current approach is better

File: Problem1.rb (hunk 13 above)
Comment (MateaAndrei) [reply Problem1.rb:4]: Yes i know, but i haven't use it. I will format my entire code for my next commit.
